test/alternator/test_returnvalues.py

```diff
@@ -83,6 +83,47 @@ def test_put_item_returnvalues(test_table_s):
     with pytest.raises(ClientError, match='ValidationException'):
         test_table_s.put_item(Item={'p': p, 'a': 'hello'}, ReturnValues='none')
 
+def skip_if_returnvalues_on_condition_check_not_supported():
+    import botocore
+    from packaging.version import Version
+    # This release added support for ReturnValuesOnConditionCheckFailure
+    if (Version(botocore.__version__) < Version('1.29.164')):
+        pytest.skip("Botocore version 1.29.164 or above required to run this test")
+
+# Testing ReturnValuesOnConditionCheckFailure feature which returns values only
+# on failed condition expression.
+def test_put_item_returnvalues_on_condition_check_failure(test_table_s):
+    skip_if_returnvalues_on_condition_check_not_supported()
+    p = random_string()
+    test_table_s.put_item(Item={'p': p, 's': 'dog'})
+    # Succesful conditional put doesn't return values.
+    ret=test_table_s.put_item(Item={'p': p, 's': 'cat'},
+        ReturnValuesOnConditionCheckFailure='ALL_OLD',
+        ConditionExpression='s = :v1',
+        ExpressionAttributeValues={':v1' : 'dog'})
+    assert not 'Attributes' in ret # Ordinary field not present.
+    assert not 'Item' in ret # Field used for error responses also not present.
+    # Failed conditional put returns old values.
+    with pytest.raises(test_table_s.meta.client.exceptions.ConditionalCheckFailedException) as err:
+        test_table_s.put_item(Item={'p': p, 's': 'cow'},
+            ReturnValuesOnConditionCheckFailure='ALL_OLD',
+            ConditionExpression='s = :v1',
+            ExpressionAttributeValues={':v1' : 'dog'})
+    assert err.value.response['Item'] == {'p': {'S': p}, 's': {'S': 'cat'}}
+    # Failed conditional put without returning old values.
+    with pytest.raises(test_table_s.meta.client.exceptions.ConditionalCheckFailedException) as err:
+        test_table_s.put_item(Item={'p': p, 's': 'cow'},
+            ReturnValuesOnConditionCheckFailure='NONE',
+            ConditionExpression='s = :v1',
+            ExpressionAttributeValues={':v1' : 'dog'})
+    assert not 'Item' in err.value.response
+    # Invalid ReturnValuesOnConditionCheckFailure value returns error.
+    with pytest.raises(ClientError, match='ValidationException'):
+        test_table_s.put_item(Item={'p': p, 's': 'cow'},
+            ReturnValuesOnConditionCheckFailure='DUMMY',
```

Comment:
Thanks for adding this. Maybe you can also check (a loop?) a bunch of wrong names, including the names ALL_NEW etc. which are legal for ReturnValues but not for  ReturnValuesOnConditionCheckFailure.